File: frontend/types/prediction.ts
/** Mirrors backend/app/schemas/prediction.py's TopKEntry. */
export interface TopKEntry {
  class: string;
  confidence: number;
}

/** Mirrors backend/app/schemas/prediction.py's BoundingBoxSchema. */
export interface BoundingBox {
  x_min: number;
  y_min: number;
  x_max: number;
  y_max: number;
  confidence: number;
}

/** Mirrors backend/app/schemas/prediction.py's PredictionResponse. */
export interface Prediction {
  id: string;
  predicted_class: string;
  confidence: number;
  top_k: TopKEntry[];
  source: "upload" | "webcam";
  latency_ms: number;
  image_path: string | null;
  bounding_box: BoundingBox | null;
  /** true = hand found, false = detector ran but found nothing (full-frame
   * fallback used), null = hand detection wasn't available at all. */
  hand_detected: boolean | null;
  created_at: string | null;
}

/** Mirrors backend/app/schemas/metrics.py's HealthResponse. */
export interface HealthStatus {
  status: string;
  model_loaded: boolean;
  model_architecture: string | null;
  device: string | null;
}

/** Mirrors backend/app/schemas/metrics.py's LetterCount. */
export interface LetterCount {
  letter: string;
  count: number;
}

/** Mirrors backend/app/schemas/metrics.py's MetricsResponse. */
export interface Metrics {
  total_predictions: number;
  average_confidence: number;
  average_latency_ms: number;
  most_predicted_letters: LetterCount[];
  predictions_by_source: Record<string, number>;
}

export interface HistoryFilters {
  limit?: number;
  offset?: number;
  source?: "upload" | "webcam";
  predicted_class?: string;
}

File: backend/app/schemas/prediction.py
"""schemas/prediction.py -- request/response contracts for the /predict* endpoints."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class TopKEntry(BaseModel):
    class_: str = Field(alias="class")
    confidence: float

    model_config = {"populate_by_name": True}


class BoundingBoxSchema(BaseModel):
    """Normalized (0-1) hand bounding box -- fraction of frame width/height,
    so the frontend can draw it correctly at any displayed video size."""

    x_min: float
    y_min: float
    x_max: float
    y_max: float
    confidence: float


class PredictionResponse(BaseModel):
    id: str
    predicted_class: str
    confidence: float
    top_k: list[TopKEntry]
    source: str
    latency_ms: float
    image_path: str | None = None
    bounding_box: BoundingBoxSchema | None = None
    # True=found, False=ran but found none (full-frame fallback used), None=detector unavailable
    hand_detected: bool | None = None
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


class WebcamFrameRequest(BaseModel):
    """The frontend sends webcam frames as a base64-encoded JPEG/PNG data
    URL (e.g. from HTMLCanvasElement.toDataURL()) rather than multipart
    file upload, since that's the natural format coming out of a <canvas>
    snapshot of a <video> element."""

    image_base64: str = Field(..., description="Base64-encoded image, with or without the data: URL prefix")
